compiler/rustc_codegen_ssa/src/mir/rvalue.rs

```diff
@@ -819,8 +819,14 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     .builtin_deref(true)
                     .unwrap_or_else(|| bug!("deref of non-pointer {:?}", input_ty))
                     .ty;
-                let llty = bx.cx().backend_type(bx.cx().layout_of(pointee_type));
-                bx.inbounds_gep(llty, lhs, &[rhs])
+                let pointee_layout = bx.cx().layout_of(pointee_type);
+                if pointee_layout.is_zst() {
+                    // ZSTs are weird, since it's legal to `.add(usize::MAX)` on them.
```

Comment:
Tbh I'd prefer a comment like "offset works in terms of the size of pointee, so offsetting a pointer to ZST is a noop", since the one here doesn't really explain why this just returns `lhs`.